### PR TITLE
Birthday field feature #1

### PR DESCRIPTION
**Pull request by justinpoh**

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Birthday.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    /**
+     * Returns true if the date is a leap year day.
+     * A leap year day must be on 29 April.
+     */
+    private static boolean isLeapYearDate(int[] processedSplitDate) {
+        if (processedSplitDate[DATE_DAY_INDEX] == LEAP_YEAR_DAY) {
+            if (processedSplitDate[DATE_MONTH_INDEX] == LEAP_YEAR_MONTH_FEBRUARY) {
+                if (isLeapYear(processedSplitDate[DATE_YEAR_INDEX])) {
+                    return true;
+                }
```
Comment (chuaweiwen) [inline src/main/java/seedu/address/model/person/Birthday.java]: Avoid this "arrow" pattern with nested-if. You can just combine the boolean expressions:
```
if (processedSplitDate[DATE_DAY_INDEX] == LEAP_YEAR_DAY
    && processedSplitDate[DATE_MONTH_INDEX] == LEAP_YEAR_MONTH_FEBRUARY
    && isLeapYear(processedSplitDate[DATE_YEAR_INDEX]))
    return true;
}
```

Comment (justinpoh) [reply src/main/java/seedu/address/model/person/Birthday.java]: But wouldnt combining be longer? Like it becomes hard to read

File: src/main/java/seedu/address/model/person/Birthday.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+/**
+ * Represents a Person's birthday in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidBirthday(String)}
+ */
+public class Birthday {
+
+    public static final String MESSAGE_BIRTHDAY_CONSTRAINTS = "Birthday must be in the following format: "
+            + "dd/mm/yyyy, dd.mm.yyyy or dd-mm-yyyy.\nLeading zeroes are allowed for the day and month field. "
+            + "The year field must have 4 digits.\nExample: 21/10/1995, 21-05-1996. 8.10.1987";
+    private static final int[] MONTH_TO_DAY_MAPPING = {31, 28, 31, 30, 31, 30, 31, 31,
```
Comment (chuaweiwen) [inline src/main/java/seedu/address/model/person/Birthday.java]: Putting the sentence after `\n` to a new line would be better?

File: src/main/java/seedu/address/model/person/Birthday.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    /**
+     * Returns true if the date is valid. A date is valid if it is possible for it to exist.
+     */
+    private static boolean isValidDate(int[] processedSplitDate) {
+        requireNonNull(processedSplitDate);
+        if (isLeapYearDate(processedSplitDate)) {
+            return true;
+        }
+        if (isYearValid(processedSplitDate[DATE_YEAR_INDEX])
+                && isDayAndMonthValid(processedSplitDate[DATE_DAY_INDEX], processedSplitDate[DATE_MONTH_INDEX])) {
```
Comment (chuaweiwen) [inline src/main/java/seedu/address/model/person/Birthday.java]: If this is too long, maybe you can put it like this?
```
if (isYearValid(processedSplitDate[DATE_YEAR_INDEX])
    && isDayAndMonthValid(processedSplitDate[DATE_DAY_INDEX],
    processedSplitDate[DATE_MONTH_INDEX])) {
    return true;
}
```

File: src/main/java/seedu/address/model/person/Birthday.java (hunk 2 above)
Comment (justinpoh) [reply src/main/java/seedu/address/model/person/Birthday.java]: Yes will change that

File: src/main/java/seedu/address/model/person/Birthday.java (hunk 3 above)
Comment (justinpoh) [reply src/main/java/seedu/address/model/person/Birthday.java]: Actually I will just follow the first one you suggested. Seems better.

File: src/main/java/seedu/address/model/person/Birthday.java (hunk 1 above)
Comment (chuaweiwen) [reply src/main/java/seedu/address/model/person/Birthday.java]: It is still doing the same thing, so nested-if isn't necessary. It still has the same number of lines and I can still read the expressions easily.

File: src/main/java/seedu/address/model/person/Birthday.java (hunk 3 above)
Comment (justinpoh) [reply src/main/java/seedu/address/model/person/Birthday.java]: idt this one should split. Its not too long and splitting in the middle of arguments is a bit confusing

**Comment by justinpoh** [comment]
There is a test in EditCommandSystemTest where the state of model is changed throughout the test. So when I added in a new test to test the clearing of the birthday field, I had to restore the model to its previous state first. Anyone have any idea how to do this in a better way?

**Comment by chuaweiwen** [comment]
@justinpoh Just create a new temporary model and test on it? It can hold on the previous state while you test on the another model.

**Comment by justinpoh** [comment]
@chuaweiwen the test simulates the application with some default values, so there isn't really a temporary model to create. The other way is to remove the tags from the editedPerson to synchronize.

**Pull request merged**